doc/AEPs/aep-0002.md

```diff
@@ -0,0 +1,210 @@
+    AEP: 2
+    Title: A complete develop/publish workflow for Alire crates
+    Author: Alejandro R. Mosteo <amosteo@unizar.es>
+    Status: Draft
+    Created: 06-jul-2020
+
+Abstract
+========
+
+Until date, we do not offer guidelines for a complete closed loop on how to use
+Alire to develop/publish/update a crate. Likewise, we do not have official
+guidelines on how to distribute, outside of the community index, a project that
+uses Alire to obtain its dependencies. 
+
+This document discusses a possible solution for both the closed-loop lifecycle
+of a crate, and the open-loop distribution of projects/releases that are
+themselves not (yet) indexed in Alire.
+
+Rationale
+=========
+
+The root of the problem
+-----------------------
+
+The file `${crate}/alire/crate.toml` (henceforth, the manifest file) is
+currently regenerated from index information during `alr get`, and rewritten
+when dependencies are modified via `alr with`. This is a lossy
+process (although this could be fixed) in which the manifest becomes
+platform-dependent. Hence, we advice against including this
+file in version control, and we ask for it to be excluded from sources
+submitted to the community index.
+
+Even if this file regeneration were not lossy, machine-generated output mangles the
+original formatting. Even with a good pretty-printer for TOML, any comments in
+the original file would be lost. This goes against maintainers, that have to
+track this manual file somehow, yet exclude it from submissions.
+
+On the other hand, developers publishing code outside of Alire, with
+dependencies from Alire, could distribute the manifest file in place and ask
+the users to run `alr update` to obtain all dependencies. This is currently
+frowned upon, again because this can only work today for manifests that are
+platform-independent.
+
+Proposal
+--------
+
+Given that Alire's index and operation revolves around manifest files that have
+to be manually crafted (at a minimum for submission to the community index), the
+proposal is to fully embrace this manual edition and go in Rust's Cargo
+direction. That is, make the manifest file a user-edited file only, and put it
+under version control.
+
+A problem that Cargo has not is that `alr with` is leveraged to comfortably
```

Comment:
> Cargo doesn't have any command to add a dependency, right? cargo.toml is always edited manually.

Right.

> For OPAM and Conan, looks like there is an install command but it doesn't change the "manifest".

Since we don't install, I think that goes beyond what we have right now anyway.

> So it would mean that the features provided by alr with are beyond what is commonly available. The more common scheme is to add new dependencies by editing a file.

Right again. In Cargo at least, you edit and the next cargo command detects the edition and updates dependencies as needed. So it's always a two-step process.